Add first post about PureNix (#2)
cdepillabout/open-blog-posts: 2021-12-27-getting-started-with-purenix.md

subdirectory :: FilePath -> FilePath -> FilePath
subdirectory p1 p2 = p1 <> "/" <> p2
```

The PureScript compiler compiles code like this to JavaScript.  PureScript ends
up being a Haskell-like language that you can use to write frontend code.

One of the main build tools for PureScript is
[Spago](https://github.com/purescript/spago).  Spago is somewhat similar to
Haskell's `stack` or `cabal`, Rust's `cargo`, etc.

One interesting feature of the PureScript compiler and Spago is that there is
good support for switching out the backend code generator.  The code generator
built into the main PureScript compiler is for JavaScript, but there are
alternative code generators for other languages.  PureNix is an alternative
backend code generator for Nix.

## PureNix

PureNix works together with PureScript and Spago to compile PureScript code
into Nix code.  Here's a small example of PureScript code, and the Nix code it
gets compiled to:

```purescript
module Main where

greeting :: String
greeting = "Hello, world!"

identity :: forall a. a -> a
identity a = a

data Maybe a = Nothing | Just a

fromMaybe :: forall a. a -> Maybe a -> a
fromMaybe a Nothing = a
fromMaybe _ (Just a) = a
```

> PureNix compiles this PureScript code to the following Nix code[^1]:

Reviewer: FWIW, footnotes can be named, and the body doesn't have to be at the bottom: https://github.com/jonascarpay/jonascarpay.github.io/blob/master/posts/2021-05-16-parsing-programming-language.md?plain=1#L147

Author: Oh thanks!  That's much nicer!  I fixed this in 78c9dd8.

      __patternFail = builtins.throw "Pattern match failure in src/Main.purs at 11:1 - 11:41";
    in
      __pattern0 (__pattern1 __patternFail);
  identity = v: v;
in
  {inherit greeting Nothing Just fromMaybe;}
```

Except for the pattern matching in `fromMaybe`, this should look pretty
straight-forward.  You can see the PureScript string `greeting` becomes a
simple Nix string.  `identity` has become a Nix function.  The constructors for
`Maybe` (`Nothing` and `Just`) have become Nix functions[^2].  `fromMaybe` has
become a Nix function.  If you work through the `__patternX` continuations in
`fromMaybe`, you can see how they
[emulate the pattern matching](https://jonascarpay.com/posts/2021-11-08-nix-adts.html)
in the PureScript `fromMaybe` function.

Calling functions written in PureScript from Nix is quite fun.  It feels quite
similar to working in the Haskell or PureScript REPL.  Here is an example of
using the above functions from the Nix REPL:

```console
$ nix repl ./output/Main/default.nix
nix-repl> identity greeting
"Hello, world!"
nix-repl> fromMaybe "bye" (Just "hello")
"hello"
nix-repl> fromMaybe "bye" Nothing
"bye"

> ```

Reviewer: I keep being surprised by how nice this works out. Maybe we should make a screencast out of this or something

Author: Yeah, we really should.  Aside from type-classes, it is super nice.

Now that you have a taste of PureNix, the next section explains how to get
started actually writing PureNix code.

## Getting Started

There is a post on the NixOS Discourse that explains how to

> [put together a development environment for PureNix](https://discourse.nixos.org/t/purenix-nix-backend-for-purescript/15756/3).

Reviewer: This may be worth inlining? Or making into a separate post in the series?

Author: Yeah, that's a good point.  I was basically doing the laziest thing I could here.

I thought about this a little bit, and I think what would be ideal is to have the content from https://discourse.nixos.org/t/purenix-nix-backend-for-purescript/15756/3 as an official document in the purenix repo.

I'll clean it up that post a bit and send it as a PR to the purenix repo.

Author: I've sent a PR for this as https://github.com/purenix-org/purenix/pull/38.

Author: Fixed in this blog post in 34c2a58.

    for which PureScript standard libraries still need to be ported to PureNix.
    If you're looking for a way to help out PureNix, this is a good place
    to start!
-   An issue for putting together a real
    [PureNix package set](https://github.com/purenix-org/purenix/issues/36).
    A PureScript/PureNix package set is similar to a Stackage resolver.  It is
    used by Spago to make it easier to find library versions that work together.

You may also be interested in [Pursuit](https://pursuit.purescript.org/).
This is like Hoogle, but for PureScript.  Pursuit doesn't know anything
about PureNix, but the API of the PureNix libraries that have been ported so
far is almost the same as the PureScript version of the libraries, so searching
for functions on Pursuit is still helpful.

Now that you have an idea how to get started with PureNix, what sorts of
developers will find PureNix easy to use?

## Who is PureNix for?

If you have experience with PureScript, Nix, and Haskell, PureNix will be easy
to get started with.  You may have to write some FFI for Nix builtins, but
other than that you should have no trouble using PureNix.  It should feel

> quite natural.

Reviewer: I feel like this section is a little add odds with what this post seems to be -- a practical introduction to PureNix. This seems more at home in post 2? Or maybe it's worth moving post 2 up, as a non-technical introduction?

Author: That's a good point.  Maybe I'll take this section and just make it its own post.  I could probably expand it a little as well.

Author: I removed this section in 80aa5e6.

Author: I re-added this section as a new post in 69a0505.

If you have experience with Nix (but not Haskell or PureScript), you will
likely have trouble getting started with PureNix.  I would recommend
first trying to learn Haskell or PureScript.  If your goal is to use PureNix,
then it doesn't really matter if you first learn Haskell or PureScript, since
they are so similar.  Haskell seems to have more beginner-oriented resources
than PureScript.

Keep in mind that like Nix, Haskell and PureScript have a steep learning curve.
Becoming proficient in Haskell or PureScript can be quite time-consuming.

## Conclusion

If you have experience with Haskell or PureScript, PureNix should be easy to
get started with.  If you don't have experience with Haskell or PureScript,
it may take a lot of work until you feel comfortable with PureNix.

*This post is the first post in a
[series about PureNix](./2021-12-26-purenix).*

## Using PureNix Professionally

If your company is considering PureNix in order to tame a complicated Nix
codebase, Jonas and I currently have time available consulting or freelance.
Feel free to [get in touch](/about).  We are also available for any other
Nix/Haskell/PureScript-related work.

## Footnotes

> [^1]: This has been slightly simplified to make it a little easier to understand.

Reviewer: ```suggestion
[^1]: The formatting has been slightly cleaned up to make it a little easier to understand.
```

Author: Thanks, fixed in add2b8b.
